binaries/build_wheels.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A script that downloads ffmpeg, ffprobe, and packager static builds for all
+the platforms we build for and then builds distribution wheels for them.
+"""
+
+import os
+import sys
+import urllib.request
+import setuptools # type: ignore
+import setuptools.command.build_py #type: ignore
+import streamer_binaries
+
+
+# Version constants.
+# Change to download different versions.
+FFMPEG_VERSION = 'n4.4-1'
+PACKAGER_VERSION = 'v2.99.5'
+
+# A map of postfixes that will be combined with the binary download links
+# to achieve a full download link.  Different postfix for each platform.
+# Extend this dictionary to add more platforms.
+PLATFORM_POSTFIXES = {
+    # 64-bit Windows
+    'win_amd64': '-win-x64.exe',
+    # 64-bit Linux
+    'manylinux1_x86_64': '-linux-x64',
+    # Linux on ARM
+    'manylinux2014_aarch64': '-linux-arm64',
+    # 64-bit with 10.9 SDK
+    'macosx_10_9_x86_64': '-osx-x64',
+  }
+
+FFMPEG_DL_PREFIX = 'https://github.com/joeyparrish/static-ffmpeg-binaries/releases/download/' + FFMPEG_VERSION
+PACKAGER_DL_PREFIX = 'https://github.com/joeyparrish/shaka-packager/releases/download/' + PACKAGER_VERSION
+
+# The download links to each binary.  These download links
+# aren't complete, they miss the platfrom-specific postfix.
+BINARIES_DL = [
+    FFMPEG_DL_PREFIX + '/ffmpeg',
+    FFMPEG_DL_PREFIX + '/ffprobe',
+    PACKAGER_DL_PREFIX + '/packager',
+  ]
+
+
+class custom_build_py(setuptools.command.build_py.build_py):
+  """A custom class to override the default behavior of `build_py` command."""
+
+  platform_build_lib = ''
+
+  def initialize_options(self):
+    return_val = super().initialize_options()
+    # Sets the `--build-lib` directory.
+    self.build_lib = custom_build_py.platform_build_lib
+    return return_val
+
+
+def build_bdist_wheel(platform_name, platform_binaries):
+  """Builds a wheel distribution for `platform_name` adding the files
+  in `platform_binaries` to it using the `package_data` parameter."""
```

Comment:
So the `platform_binaries` argument is how you are avoiding using a temp directory?  You are allow-listing the binaries per-platform in each call to build a wheel?

Reply:
ummm... not exactly, 
platform_binaries is used to tell `setuptools` which files in the package(`streamer_binaries`) to include. By default, `setuptools` will include all the `.py` files and leave the others, will not put them into the wheel, by specifying them as `package_data`, we are telling `setuptools` to ship them into the `.whl` file it is going to generate with all the `.py` files.

The build directory problem is that when `setuptools` is building the package for one platform, it dumps every thing it will ship in `build/lib`(the default), then it doesn't delete it, the next platform in the for loop will be built in the same directory `build/lib`, but some old binaries are still there, thus we have to clean this directory after compressing the wheel for each platform.

We can't set `--build-lib` from the `bdist_wheel`. `bdist_wheel` calls `build`, and `build` has the `--build-lib` option, so i set the from the `build` command directly. By setting each build directory to the platform name we avoid reusing the same build directory so we are fine :D.